Data/ByteString/Short/Internal.hs

```diff
@@ -1890,3 +2005,29 @@ moduleError :: HasCallStack => String -> String -> a
 moduleError fun msg = error (moduleErrorMsg fun msg)
 {-# NOINLINE moduleError #-}
 
+------------------------------------------------------------------------
+-- IO
+
+-- | Outputs 'ShortByteString' to the specified 'Handle'. This is implemented
+-- with 'IO.hPutBuf'. The offset and length are used because we don't have slices of 'ByteArray' yet.
+-- The function is unsafe because the offset and length is not checked.
+unsafeHPutOff :: IO.Handle -> ShortByteString -> Int -> Int -> IO ()
+unsafeHPutOff handle sbs off len = withPtr sbs $ \p -> IO.hPutBuf handle (p `plusPtr` off) len
+
+hPut :: IO.Handle -> ShortByteString -> IO ()
+hPut h sbs = unsafeHPutOff h sbs 0 (length sbs)
```

Comment:
There's no need to re-invent the wheel here. `fromShort` already basically does what the pointer-related conversions above are meant to.

```suggestion
hPut h sbs = BS.hPut h (fromShort sbs)
```
To keep from allocating the intermediate `ByteString`, we can just move the pinnedness check from `fromShort` into `fromShortIO` and then use the latter.
```suggestion
hPut h sbs = fromShortIO sbs >>= BS.hPut h
```